cachix/src/Cachix/Client/Commands.hs

```diff
@@ -205,125 +203,52 @@ push env name _ True = withManager $ \mgr -> do
-                let deriver = if deriverRaw == "unknown-deriver"
-                              then deriverRaw
-                              else T.drop 11 deriverRaw
-
-                references <- sort . T.lines . T.strip . toS <$> readProcess "nix-store" ["-q", "--references", toS storePath] mempty
-
-                let
-                    fp = fingerprint storePath narHash narSize references
-                    sig = dsign sk fp
-                    nic = Api.NarInfoCreate
-                      { cStoreHash = storeHash
-                      , cStoreSuffix = storeSuffix
-                      , cNarHash = narHash
-                      , cNarSize = narSize
-                      , cFileSize = fileSize
-                      , cFileHash = fileHash
-                      , cReferences = fmap (T.drop 11) references
-                      , cDeriver = deriver
-                      , cSig = toS $ B64.encode $ unSignature sig
-                      }
-
-                escalate $ Api.isNarInfoCreateValid nic
-
-                -- Upload narinfo with signature
-                escalate <=< (`runClientM` clientenv env) $ Api.createNarinfo
-                  (cachixBCClient name)
-                  (Api.NarInfoC storeHash)
-                  nic
-
-
--- Utils
-
--- Retry up to 5 times for each store path.
--- Catches all exceptions except skipAsyncExceptions
-retryPath :: (RetryStatus -> IO a) -> IO a
-retryPath = recoverAll defaultRetryPolicy
-
-defaultRetryPolicy :: RetryPolicy
-defaultRetryPolicy =
-  constantDelay 50000 <> limitRetries 3
-
-mapConcurrentlyBounded :: Traversable t => Int -> (a -> IO b) -> t a -> IO (t b)
-mapConcurrentlyBounded bound action items = do
-  qs <- QSem.newQSem bound
-  let wrapped x = bracket_ (QSem.waitQSem qs) (QSem.signalQSem qs) (action x)
-  mapConcurrently wrapped items
-
-splitStorePath :: Text -> (Text, Text)
-splitStorePath storePath =
-  (T.take 32 (T.drop 11 storePath), T.drop 44 storePath)
+  escalateAs FatalError $ parseSigningKeyLenient $ fromJust $ maybeBCSK <|> toS <$> maybeEnvSK <|> panic "You need to: export CACHIX_SIGNING_KEY=XXX"
+
+retryText :: RetryStatus -> Text
+retryText retrystatus =
+  if rsIterNumber retrystatus == 0
+  then ""
+  else "(retry #" <> show (rsIterNumber retrystatus) <> ") "
+
+pushStrategy :: Env -> Text -> Text -> PushStrategy IO ()
+pushStrategy env name storePath = PushStrategy
+  { onAlreadyPresent = pass
+  , on401 = if isJust (config env)
+            then throwM $ accessDeniedBinaryCache name
+            else throwM notAuthenticatedBinaryCache
+  , onError = throwM
+  , onAttempt = \retrystatus -> 
+      -- we append newline instead of putStrLn due to https://github.com/haskell/text/issues/242
+      putStr $ "pushing " <> retryText retrystatus <> storePath <> "\n"
+  , onDone = pass
+  , withXzipCompressor = defaultWithXzipCompressor
+  }
+
+pushStorePath :: Env -> Text -> Text -> IO ()
```

Comment:
This is preserved to keep Haskell API stable?

Reply:
Yes, and it's also used by `--watch-store`.